### DISCUSSION TITLE
ERROR:  invalid transaction termination

### DISCUSSION THREAD

**Discussion by khushboovashi** [psycopg/psycopg#797] [General]
I have two procedures as below:

```
CREATE OR REPLACE PROCEDURE public.test1_proc(
	)
LANGUAGE 'plpgsql'
AS $BODY$
DECLARE 
	r RECORD;
BEGIN
	FOR r IN SELECT * FROM (values(1,'test1'),(2,'test2')) as c(id,name) LOOP
		INSERT INTO test_table VALUES (r.id, r.name);
		COMMIT;
	END LOOP;
END;
$BODY$;
```

```
CREATE OR REPLACE PROCEDURE public.test2_proc(
	)
LANGUAGE 'plpgsql'
AS $BODY$
DECLARE 
	r RECORD;
BEGIN
	FOR r IN SELECT * FROM (values(3,'test1'),(4,'test2')) as c(id,name) LOOP
		INSERT INTO test_table VALUES (r.id, r.name);
		COMMIT;
	END LOOP;
END;
$BODY$;
```

If I execute them individually, they work fine, but if I try to call them in a single query like` _cur.execute('CALL test1_proc(); CALL test2_proc();'),` it fails with the below error.

```
ERROR:  invalid transaction termination
CONTEXT:  PL/pgSQL function test1_proc() line 7 at COMMIT 

SQL state: 2D000
```

Can someone explain the reason for that?

**Comment by dvarrazzo** [comment]
It is explained [here](https://www.psycopg.org/psycopg3/docs/basic/from_pg2.html#multiple-statements-in-the-same-query), see the warning block.

> If a statement must be executed outside a transaction (such as CREATE DATABASE), it cannot be executed in batch with other statements, even if the connection is in autocommit mode:
>
> ```python
> conn.autocommit = True
>
> conn.execute("CREATE DATABASE foo; SELECT 1")
> Traceback (most recent call last):
> ...
> psycopg.errors.ActiveSqlTransaction: CREATE DATABASE cannot run inside a transaction block
> ```
>
> This happens because PostgreSQL itself will wrap multiple statements in a transaction. Note that you will experience a different behaviour in psql (psql will split the queries on semicolons and send them to the server separately).
>
> This is not new in Psycopg 3: the same limitation is present in psycopg2 too.
